lib/Conversion/XTenNNToTosa.cpp

```diff
@@ -0,0 +1,226 @@
+  IntegerType intType = dyn_cast<IntegerType>(type.getElementType());
+  if (!intType) {
+    return IntegerMinMax{};
+  }
+  auto minValue =
+      llvm::APSInt::getMinValue(intType.getWidth(), type.isUnsignedInteger());
+  auto maxValue =
+      llvm::APSInt::getMaxValue(intType.getWidth(), type.isUnsignedInteger());
+  return IntegerMinMax{minValue.getSExtValue(), maxValue.getSExtValue()};
+}
+
+class QuantizeOp : public OpRewritePattern<amd::xten_nn::QuantizeOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(amd::xten_nn::QuantizeOp quantizeOp,
+                                PatternRewriter &rewriter) const override {
+    // The QDQ operations only work on tensors, if they are not, then the
+    // verifiers should find the error. At the moment, only the signed tensors
+    // are supported.
+    auto outputType = cast<TensorType>(quantizeOp->getResult(0).getType());
+    if (!outputType.getElementType().isSignedInteger()) {
+      return rewriter.notifyMatchFailure(
+          quantizeOp.getLoc(), "only signed tensor types are supported.");
+    }
+    auto inputType = dyn_cast<TensorType>(quantizeOp->getOperand(0).getType());
+    // TOSA only supports signed integers of i8, i16 or i32 here we convert our
+    // si<?> to this types and add a clamp to mimic arbitrary bit width.
+    TensorType newIntegerStorageType = getNewStorageType(outputType);
+
+    // Calculate (1 / 2 ^ shift )
+    llvm::APFloat scale(std::pow(static_cast<float>(2.0),
+                                 static_cast<float>(-quantizeOp.getShift())));
+
+    // Create a constant that represents the (1 / 2 ^ shift )
+    RankedTensorType constType =
+        createSplatType(inputType.getRank(), rewriter.getF32Type());
+    auto constOp = rewriter.create<tosa::ConstOp>(
+        quantizeOp->getLoc(), constType,
+        DenseFPElementsAttr::get(constType, {scale}));
+
+    // Calculate ( x / 2 ^ shift)
+    auto mulOp = rewriter.create<tosa::MulOp>(
+        quantizeOp.getLoc(), inputType, quantizeOp->getOperand(0),
+        constOp->getResult(0), rewriter.getI32IntegerAttr(0));
+
+    // Cast from fp32 -> i<Bitwidth> where bit width is the supported storage
+    // bit width. Either i8, i16 or i32
+    auto castOp = rewriter.create<tosa::CastOp>(
+        quantizeOp->getLoc(), newIntegerStorageType, mulOp->getResult(0));
+
+    // Find the max and min of the signed integer type.
+    IntegerMinMax intLimits = calculateMinMaxOfElementType(outputType);
+
+    // Clamp the integer to the min and max we calculated for this custom
+    // bit width
+    auto clampOp = rewriter.createOrFold<tosa::ClampOp>(
+        quantizeOp->getLoc(), newIntegerStorageType, castOp->getResult(0),
+        rewriter.getI64IntegerAttr(intLimits.min),
+        rewriter.getI64IntegerAttr(intLimits.max),
+        rewriter.getF32FloatAttr((float)intLimits.min),
+        rewriter.getF32FloatAttr((float)intLimits.max));
```

Comment:
This is an integer clamp, right? The computation for min/max float values here might result in imprecise values (for large integer limits) and the values should be unused (I think), so maybe simply put dummy values for float min/max instead.
We should probably propose some different way how to represent int/float clamps, really weird to always supply both int and float min/max.

Reply:
Could we consider what's in the clamp for floats already dummy? I understand for integers > 24 bits we cannot represent them in fp32 1:1 but the attributes need some value for the moment.